### PR TITLE
Add joinaggregate. Fixes #4121

### LINKED ISSUE #4121: Join aggregate transform
We already support window but joinaggregates are probably the most common and are supported natively in Vega. 

See #3912

### PR DESCRIPTION
**Pull request by domoritz**

### CODE REVIEW

File: src/transform.ts [hunk 1]
```diff
@@ -207,6 +208,35 @@ export interface WindowTransform {
...
 }
 
+export interface JoinAggregateDef {
+  /**
+   * The aggregation operation to apply (e.g., sum, average or count). See the list of all supported operations [here](https://vega.github.io/vega-lite/docs/aggregate.html#ops).
+   */
+  op: AggregateOp;
+
+  /**
+   * The data field for which to compute the aggregate function. This can be omitted for window functions that do not operate over a field such as `count`.
```
Comment (kanitw) [inline src/transform.ts]: >  for window functions

This no longer "window functions" right?

File: .prettierignore [hunk 2]
```diff
@@ -0,0 +1 @@
+examples/specs/*.vl.json
```
Comment (kanitw) [inline .prettierignore]: Can we put this rule in `.prettierrc.json`?

If we can, probably better to consolidate them in one place.

File: src/transform.ts (hunk 1 above)
Comment (domoritz) [reply src/transform.ts]: Fixed

File: .prettierignore (hunk 2 above)
Comment (domoritz) [reply .prettierignore]: I don't think it works yet. https://github.com/prettier/prettier/issues/3460

File: examples/compiled/area_cumulative_freq.vg.json [hunk 3]
```diff
@@ -12,13 +12,10 @@
       "format": {"type": "json"},
       "transform": [
         {
-          "type": "window",
-          "params": [null],
+          "type": "joinaggregate",
```
Comment (kanitw) [inline examples/compiled/area_cumulative_freq.vg.json]: Cumulative frequency should not use join aggregate.  

Window with frame = `[null, 0]` makes more sense. (See SVG)

File: examples/compiled/layer_cumulative_histogram.vg.json [hunk 4]
```diff
@@ -33,13 +33,10 @@
...
         {
-          "type": "window",
-          "params": [null],
+          "type": "joinaggregate",
           "as": ["cumulative_count"],
           "ops": ["sum"],
-          "fields": ["count"],
-          "sort": {"field": ["bin_IMDB_Rating"], "order": ["ascending"]},
-          "frame": [null, 0]
+          "fields": ["count"]
```
Comment (kanitw) [inline examples/compiled/layer_cumulative_histogram.vg.json]: Cumulative frequency should not use join aggregate.  

Window with frame = `[null, 0]` makes more sense. (See SVG)

File: examples/compiled/waterfall_chart.vg.json [hunk 5]
```diff
@@ -30,12 +30,10 @@
       "source": "source_0",
       "transform": [
         {
-          "type": "window",
-          "params": [null],
+          "type": "joinaggregate",
```
Comment (kanitw) [inline examples/compiled/waterfall_chart.vg.json]: 1) Don't rename this file

2) waterfall should use window (with default frame = `[null, 0]`)

File: examples/compiled/window_cumulative_running_average.vg.json [hunk 6]
```diff
@@ -19,14 +19,10 @@
           "expr": "datetime(year(datum[\"Year\"]), 0, 1, 0, 0, 0, 0)"
         },
         {
-          "type": "window",
-          "params": [null],
+          "type": "joinaggregate",
```
Comment (kanitw) [inline examples/compiled/window_cumulative_running_average.vg.json]: wrong

File: examples/specs/bar_diverging_stack_transform.vl.json [hunk 7]
```diff
@@ -63,8 +63,8 @@
           "as": "offset"
         }
       ],
-      "groupby": ["question"],
-      "frame": [null, null]
+      "frame": [null, null],
```
Comment (kanitw) [inline examples/specs/bar_diverging_stack_transform.vl.json]: This one can be joinaggregate

File: examples/compiled/area_cumulative_freq.vg.json (hunk 3 above)
Comment (domoritz) [reply examples/compiled/area_cumulative_freq.vg.json]: Ups, my mistake.

Comment (domoritz) [reply examples/compiled/area_cumulative_freq.vg.json]: The code should only do it when the frame is `[null, null]`

Comment (domoritz) [reply examples/compiled/area_cumulative_freq.vg.json]: This was a bug in the code. Fixed.

File: examples/specs/waterfall_chart.vl.json [hunk 8]
```diff
@@ -21,7 +21,7 @@
   "width": 800,
   "height": 450,
   "transform": [
-    {"window": [{"op": "sum", "field": "amount", "as": "sum"}]},
+    {"joinaggregate": [{"op": "sum", "field": "amount", "as": "sum"}]},
```
Comment (domoritz) [inline examples/specs/waterfall_chart.vl.json]: @kanitw We use joinaggregate for the window where it makes sense. The name became a bit unwieldy and I don't think we need to list every feature in the filename for this complex example. I was thinking that we maybe want a separate section for complex/custom designs at the bottom of the examples page.

Comment (kanitw) [reply examples/specs/waterfall_chart.vl.json]: You can keep the name if you have strong preference, but window doesn't make sense here.   See SVG.

**Comment by kanitw** [comment]
Instead of just doing it in the code, I think better update all the examples (except one of them) to use joinaggregate in VL examples too.

**Comment by domoritz** [comment]
> Instead of just doing it in the code, I think better update all the examples (except one of them) to use joinaggregate in VL examples too.

That's what I did. Which ones did I miss?

**Comment by domoritz** [comment]
Fixed bar_diverging as well

**Comment by domoritz** [comment]
@kanitw This is ready

File: examples/compiled/waterfall_chart.vg.json (hunk 5 above)
Comment (kanitw) [reply examples/compiled/waterfall_chart.vg.json]: Waterfall chart is still wrong.  It should not use `joinaggregate`.

Comment (kanitw) [reply examples/compiled/waterfall_chart.vg.json]: image

File: site/docs/transform/joinaggregate.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
+---
+layout: docs
+menu: docs
+title: Join Aggregate
+permalink: /docs/joinaggregate.html
+---
+
+The joinaggregate transform extends the input data objects with aggregate values. Aggregation is performed and the results are then joined with the input data. This transform can be helpful for creating derived values that combine both raw data and aggregate calculations, such as percentages of group totals. This transform is a special case of the [window](window.html) transform where the `frame` is always `[null, null]`.
```
Comment (kanitw) [inline site/docs/transform/joinaggregate.md]: > The joinaggregate transform extends the input data objects with aggregate values.

1) Explicitly mention that the extension is a new column?  (Right now it's still a bit ambiguous.)


> This transform is a special case of the [window](window.html) transform where the `frame` is always `[null, null]`.

It's good to compare this to `window`, but perhaps it's more important to highlight the difference between this and normal aggregate.  (Similarly, add docs to aggregate to mention that "use joinaggregate if you want to preserve the table structure and instead add a new column for the aggregated values.

File: src/transform.ts [hunk 10]
```diff
@@ -207,6 +208,35 @@ export interface WindowTransform {
   sort?: SortField[];
 }
 
+export interface JoinAggregateDef {
```
Comment (kanitw) [inline src/transform.ts]: FWIW, the similar thing for Aggregate is called `AggregateFieldDef`, so maybe `JoinAggregateFieldDef` for consistency?

File: site/docs/transform/joinaggregate.md (hunk 9 above)
Comment (domoritz) [reply site/docs/transform/joinaggregate.md]: Done

File: src/transform.ts (hunk 10 above)
Comment (domoritz) [reply src/transform.ts]: Done

**Pull request merged**